Avoid frequent allocations

erodozer opened the pull request:
Decreases the number of objects being allocated by building the mesh upfront and performing in-place updates.

This is a reduction of #121 to address specifically issue #122

erodozer committed:
avoid frequent allocations

Significantly decrease the number of objects being allocated.  Optimizes by performing in-place updates on resources after initially building the mesh.

MizunagiKB reviewed (commented):
The `clear` function has been removed from the public API. 
Could you please explain the reason for removing a function that was already present and documented in the reference manual?
> This seems to be related to the reset item of the `Assets` property.

Since this operation pertains to the `assets` property, the `clear` function, which might appear to serve a similar purpose but could lead to different operations, has been removed in favor of focusing on "resetting the property value in the editor" and "`assets = ''`".

If this is the case, I understand the rationale.

In that case, I will update the reference manual to align with the implementation.

MizunagiKB commented on src/gd_cubism_user_model.cpp:
```diff
@@ -815,6 +763,73 @@ void GDCubismUserModel::_get_property_list(List<godot::PropertyInfo> *p_list) {
+    this->set_size(Vector2i(
+        model->GetCanvasWidthPixel(),
+        model->GetCanvasHeightPixel()
+    ));
```
Many Live2D models are high-resolution, and the default resolution of `SubViewport` at 512x512 pixels is insufficient. I assume you adjusted the `SubViewport` size to accommodate this.

However, this change is not compatible with the original behavior. The size specification for `SubViewport` should be determined by the user, so please avoid automatically altering the specified values.

That said, no one wants poor rendering due to `SubViewport` when using high-resolution textures.

To resolve this, we are considering a solution that does not use `SubViewport`. Instead, we will render directly to the screen Viewport, and those who wish to use `SubViewport` can create it themselves and place the `GDCubismUserModel` under it.

Although based on a slightly older version, we already have a working solution.

> This behavior is likely what everyone expects.

https://github.com/MizunagiKB/gd_cubism/tree/feature_direct_rendering

erodozer replied:
Sorry, I accidentally left this in as it was a hack for myself seeing distortion of aspect ratio occasionally when swapping files, although that was of my own misunderstanding and now I no longer encounter what I thought was bug.

I'm looking forward to the direct rendering, as it was something else I was experimenting on doing myself when digging into this extension.

MizunagiKB replied:
Thank you for the update.

In fact, I have seen posts on platforms like X mentioning that **"the resolution of images displayed in GDCubism is poor (this issue has been resolved)."**

Although this is not currently the case, with direct rendering, it seems best to follow the behavior of `Sprite2D`, where the image is first rendered at its original size and then adjusted to the desired size by the user.

MizunagiKB commented on src/gd_cubism_user_model.cpp:
```diff
@@ -228,53 +226,16 @@ GDCubismUserModel::moc3FileFormatVersion GDCubismUserModel::csm_get_moc_version(
 
 void GDCubismUserModel::set_assets(const String assets) {
+    ERR_FAIL_COND_MSG(!assets.ends_with(".model3.json"), "Must point to Live2D model3.json file");
+    ERR_FAIL_COND_MSG(!FileAccess::file_exists(assets), "Could not find model path");
```
Due to stricter error checking, it is no longer possible to reset the state of Assets.

This process appears to target only Live2D model files and allows only replacements, which prevents the initialization of the Assets property.

> Personally, this has made debugging somewhat cumbersome.

Additionally, if the actual Live2D model file is deleted, it means that the Live2D model cannot be returned to a state where it references nothing.
The error message during reset is also somewhat inappropriate.

Could you make it possible to reset the state of Assets?

erodozer replied:
makes sense.  I've changed these to be WARN instead of act as barriers

MizunagiKB replied:
I have confirmed the fix.

The change from error to warning has been made, and the notification message is now more appropriate. Thank you.

MizunagiKB commented on src/gd_cubism_user_model.cpp:
```diff
@@ -80,8 +80,6 @@ void GDCubismUserModel::_bind_methods() {
 
     ClassDB::bind_method(D_METHOD("get_canvas_info"), &GDCubismUserModel::get_canvas_info);
 
-    ClassDB::bind_method(D_METHOD("clear"), &GDCubismUserModel::clear);
-
```
The clear function has been removed from the public API.
Could you please explain the reason for removing a function that was already present and documented in the reference manual?

> This seems to be related to the reset item of the `assets` property.

Since this operation pertains to the `assets` property, the `clear` function, which might appear to serve a similar purpose but could lead to different operations, has been removed in favor of focusing on "resetting the property value in the editor" and "`assets = ''`".

If this is the case, I understand the rationale.

In that case, I will update the reference manual to align with the implementation.

erodozer replied:
Yes, I changed it to a private method to standardize a single approach on how to change the loaded model based on setting assets.

MizunagiKB replied:
I did not realize this until you pointed it out.
There were areas that could unknowingly result in "similar but different behavior."

Both in the editor and at runtime:

- If the assets have valid values, the Live2D model is rendered.
- If the assets have invalid values, nothing happens.

This has streamlined the process, eliminating any confusion.

I will update the documentation to reflect the removal of the `clear` function.

This was very helpful.

MizunagiKB replied:
Updated Documentation in Line with Your Pull Request.
https://github.com/MizunagiKB/gd_cubism/pull/124/commits/d5ad255c4b5781ff4b285bd93de8282837cd2e67

MizunagiKB reviewed (changes requested):
I have commented under "Request changes," but I would like to know the reason for removing the function.
> (Resolved)

erodozer committed:
only warn instead of error when trying to clear assets

MizunagiKB committed:
Removal of GDCubismUserModel::clear Function

In conjunction with the removal of the GDCubismUserModel::clear function, I have updated the relevant documentation.

MizunagiKB commented:
I noticed some concerns with the `GDCubismUserModel::clear` function.

While I could have pushed the changes, I decided to write the code directly as it is brief.

**In src/gd_cubism_user_model.cpp, line 771-**

```c++
void GDCubismUserModel::clear() {
    while (this->get_child_count(true) > 0) {
        auto const& c = this->get_child(0, true);
        c->queue_free();
        this->remove_child(c);
    }

    ...
```

Although this implementation is more verbose than your original code, `GDCubismUserModel` allows users to attach various extensions with additional gimmicks (such as eye tracking and breathing motions).
These can be freely implemented using the `GDCubismEffect` class or its derived classes.
Since the original behavior was to retain these nodes, I have modified this process to maintain compatibility.

> This code will result in the removal of everything except `GDCubismEffect` and its derived classes, which is consistent with the original behavior.

**In src/gd_cubism_user_model.cpp, line 771-**

```c++
void GDCubismUserModel::clear() {
    {
        TypedArray<Node> ary_node;

        for(int32_t index = 0; index < this->get_child_count(true); index++) {
            auto const& c = this->get_child(index, true);
            if(Object::cast_to<GDCubismEffect>(c) != nullptr) continue;
            ary_node.append(c);
        }

        while(ary_node.size() > 0) {
            auto const& c = Object::cast_to<Node>(ary_node[0]);
            c->queue_free();
            this->remove_child(c);
            ary_node.remove_at(0);
        }
    }

    ...
```

erodozer committed:
clear only nodes managed by the loader and not added by the user

erodozer commented:
> `GDCubismUserModel` allows users to attach various extensions with additional gimmicks (such as eye tracking and breathing motions). These can be freely implemented using the `GDCubismEffect` class or its derived classes. Since the original behavior was to retain these nodes, I have modified this process to maintain compatibility.

Instead of crawling the tree and checking for specific node types, I've gone ahead and adjusted the clear process to purge specifically any Node that was directly spawned and managed by the renderer.  That way user added nodes, as long as they haven't been attached directly as child of a managed node, will remain.

erodozer committed:
fix disposing models

MizunagiKB commented:
Please refrain from casually introducing new implementations when code reviews and functional testing have not been completed.
In this pull request, we aim to stop the nodes from being restructured every frame, as suggested by your code.

First, we will conduct code reviews and functional testing for the following state.
> https://github.com/MizunagiKB/gd_cubism/commit/24509a9b64e998b986f25487a97a72a636632bfd

erodozer commented:
fair enough.  I had switched to MeshDataTool because in the Godot API it's written as though it's the preferred way of manipulating array meshes after initialization, but the implementation didn't have any benefits over what I wrote before.  I've gone ahead and reverted it

erodozer committed:
update godot 4.3 bindings to fix disposal bug

erodozer commented on src/private/internal_cubism_renderer_resource.cpp:
```diff
@@ -56,78 +51,109 @@ InternalCubismRendererResource::~InternalCubismRendererResource() {
-        this->ary_sub_viewport.append(viewport);
-        return viewport;
+    for (int i = 0; i < this->managed_nodes.size(); i++) {
+        if (!UtilityFunctions::is_instance_id_valid(this->managed_nodes[i])) continue;
```
Looks like is_instance_id_valid is broken in the current version of the bindings because of https://github.com/godotengine/godot-cpp/issues/1587

After updating the godot-cpp binding to the latest 4.3 tag, I can verify that this works properly for both swapping assets and disposing the instance.  Without this check, disposing will cause a crash because this child meshes may already have been freed by godot's own cleanup.

MizunagiKB replied:
Thank you for the rollback.

However, the version you rolled back to is causing crashes when the _editor_ or _runtime_ is terminated.

> https://github.com/MizunagiKB/gd_cubism/pull/124

While there has been a response to this issue, the use of the `UtilityFunctions::is_instance_id_valid` function is somewhat of a temporary workaround.

**Here are the details:**
* It is undesirable that we are unable to manage resources that we are supposed to be managing.
* The reason for this issue is that resources are being released in the destructor (`GDCubismUserModel::~GDCubismUserModel()`).
* The destructor may be called after the Godot Engine itself has terminated, which is potentially problematic.
* To resolve this, we will modify the process to handle it in the `_notification` function.

**References for the `_notification` function:**
* [Godot Engine Documentation](https://docs.godotengine.org/en/stable/classes/class_object.html#class-object-private-method-notification)
* [Example in godot-cpp](https://github.com/godotengine/godot-cpp/blob/fbbf9ec4efd8f1055d00edb8d926eef8ba4c2cce/test/src/example.cpp#L72-L76)

Although your method also prevents crashes, it leaves potential issues unresolved.
Therefore, we will address it using this method.

> We have reverted the reference to the `godot-cpp` repository back to `4.3-stable`.
> The issue was not caused by Godot but by the code I wrote. Since the function in question is no longer needed, there is no reason to reference an unstable version.

erodozer replied:
Thanks for finding that.  I'm still pretty new to Godot's c++ bindings vs leveraging GDScript for plugins, it's not always clear when certain parts of the language that you'd expect to work conflict with the actual lifecycle of how godot manages objects :/

MizunagiKB committed:
Bug fix for graceful termination

Fixed potential issues with resource release in destructor by moving the process to the _notification function.

MizunagiKB commented:
If no issues arise, we will proceed with the merge after verifying the operation.

erodozer committed:
ignore invalid mask meshes

MizunagiKB commented:
@erodozer

A new commit has been made. Could you please specify the purpose of this code?
The commit log mentions **ignore invalid mask meshes**, but could you provide the rationale for this change for the record?

> https://github.com/MizunagiKB/gd_cubism/pull/124/commits/276aabfaa3fee000ccaaefc5f88074abb0337656

----

At the moment, only you and I are participating, so it might seem unnecessary.
However, I would like to ensure that we keep thorough records, including myself.

MizunagiKB commented on src/private/internal_cubism_renderer_2d.cpp:
```diff
@@ -281,9 +286,17 @@ void InternalCubismRenderer2D::build_model(InternalCubismRendererResource &res,
                 // https://github.com/godotengine/godot/issues/89651
                 viewport->set_transparent_background(true);
 
+                masks.resize(model->GetDrawableMaskCounts()[index]);
```
Is this change intended to fix the size of the `TypedArray<MeshInstance2D> masks` at the beginning, instead of appending them individually?

If that is the purpose, since the information can be determined at the start, I believe your intention will be better conveyed to those reading the code if it is done as follows.
What do you think?

> line 271 -
```
TypedArray<MeshInstance2D> masks;
masks.resize(model->GetDrawableMaskCounts()[index]);
```

erodozer replied:
yep, making it match the size of the metadata helps keep it efficient to iterate through.  I'll move the line up.

erodozer commented:
> A new commit has been made. Could you please specify the purpose of this code? The commit log mentions **ignore invalid mask meshes**, but could you provide the rationale for this change for the record?
> 
> > [276aabf](https://github.com/MizunagiKB/gd_cubism/commit/276aabfaa3fee000ccaaefc5f88074abb0337656)

I had noticed in testing of some models my changes had caused excessive errors to be thrown every frame, breaking intended behavior.  This was because I had forgotten to carry over the logic that ignores creating mask meshes with no vertices like with other drawables.  So the commit I added filters them out the same way.

MizunagiKB commented:
Although the progress has been delayed, the following operations have been confirmed:

* Windows x86_64
* macOS Arm64
* Linux (Ubuntu) x86_64

> Although testing is being conducted on a VM, the performance on Linux has significantly improved.

erodozer committed:
move resize

MizunagiKB commented:
Thank you for the excellent pull request, implementation, and comments.
We will now merge this into the main branch.

> Live2D itself supports iOS and Android, but as it is still experimental, we have omitted testing.
